libraries/AP_IBus_Telem/AP_IBus_Telem.cpp

```diff
@@ -0,0 +1,507 @@
+   * 0x02: Sensor length (2 or 4 bytes)
+   * 0x63: Checksum low byte
+   * 0xFF: Checksum high byte
+   Checksums are 0xFFFF minus the sum of the previous bytes
+ */
+void AP_IBus_Telem::handle_type_command()
+{
+    _sensor_definition *s = get_sensor_definition(_incoming_sensor_id);
+    struct protocol_command_type_response_t {
+        uint8_t command_length;
+        uint8_t address;
+        uint8_t type;
+        uint8_t length;
+        uint16_t checksum;
+    } packet {
+        PROTOCOL_SIX_LENGTH,
+        (uint8_t)(PROTOCOL_COMMAND_TYPE | _incoming_sensor_id),
+        s->sensor_type,
+        s->sensor_length
+    };
+    populate_checksum((uint8_t*)&packet, sizeof(packet));
+    _port->write((uint8_t*)&packet, sizeof(packet));
+}
+
+/* A value query has the following format:
+   * 0x04: Message length
+   * 0xA1: 0xA0 for value + 0x01 for sensor ID 1
+   * 0x5A: Checksum low byte
+   * 0xFF: Checksum high byte
+   Checksums are 0xFFFF minus the sum of the previous bytes
+
+   To respond to a value query, we send:
+   * 0x06: Message length (or 0x08 for a 4-byte sensor)
+   * 0xA1: 0xA1 for value + 0x01 for sensor ID 1
+   * 0xD4: Value byte (value is 12,500 in this example)
+   * 0x30: Value byte
+   * 0x54: Checksum low byte
+   * 0xFE: Checksum high byte
+   Checksums are 0xFFFF minus the sum of the previous bytes
+ */
+void AP_IBus_Telem::handle_value_command()
+{
+    _sensor_definition *s = get_sensor_definition(_incoming_sensor_id);
+    const _sensor_value value = get_sensor_value(s->sensor_type);
+    if (s->sensor_length == 2) {
+        handle_2_byte_value_command(s->sensor_type, value);
+    } else {
+        handle_4_byte_value_command(s->sensor_type, value);
+    }
+}
+
+void AP_IBus_Telem::handle_2_byte_value_command(uint8_t sensor_type, _sensor_value value)
+{
+    struct protocol_command_2byte_value_response_t {
+        uint8_t command_length;
+        uint8_t address;
+        uint8_t value[2];
+        uint16_t checksum;
+    } packet {
+        PROTOCOL_SIX_LENGTH,
+        (uint8_t)(PROTOCOL_COMMAND_VALUE | _incoming_sensor_id),
+        {value.byte[0], value.byte[1]}
+    };
+    populate_checksum((uint8_t*)&packet, sizeof(packet));
+    _port->write((uint8_t*)&packet, sizeof(packet));
+}
+
+void AP_IBus_Telem::handle_4_byte_value_command(uint8_t sensor_type, _sensor_value value)
+{
+    struct protocol_command_4byte_value_response_t {
+        uint8_t command_length;
+        uint8_t address;
+        uint8_t value[4];
+        uint16_t checksum;
+    } packet {
+        PROTOCOL_EIGHT_LENGTH,
+        (uint8_t)(PROTOCOL_COMMAND_VALUE | _incoming_sensor_id),
+        {value.byte[0], value.byte[1], value.byte[2], value.byte[3]}
+    };
+    populate_checksum((uint8_t*)&packet, sizeof(packet));
+    _port->write((uint8_t*)&packet, sizeof(packet));
+}
+
+AP_IBus_Telem::_sensor_definition* AP_IBus_Telem::get_sensor_definition(uint8_t sensor_id)
+{
+    return &_sensors[sensor_id - 1];
```

Comment:
There will probably only be one caller to this with above-requested changes, and this could probably then be in-lined.